File: R/produce_csv.R
# Produce tidy CSV files of ocean-current arrows from the source shapefiles.
#
# Modern, reproducible replacement for the historical `Produce csv.R`. Uses sf
# (rgdal, sp and broom are retired) to read the curated current-arrow
# shapefiles, reproject everything to WGS84 decimal degrees (EPSG:4326), and
# write long-format CSV tables for plotting in R, Python or any GIS/Matlab
# workflow. Mirrors python/produce_csv.py.
#
# Run from the repository root:  Rscript R/produce_csv.R
#
# Schema (one row per vertex): long, lat, order, group, size, type, region
#   type   = water-mass origin, "Atlantic" (warm) or "Arctic" (cold)
#   region = "Barents Sea" or "North Atlantic"

library(sf)

# Source shapefiles -> (type, region). "warm" = Atlantic water, "cold" =
# Arctic water; harmonised to the Barents Atlantic/Arctic convention. The
# North Atlantic "_v2" files are the latest, most complete versions (EPSG:3995)
# and are reprojected to decimal degrees here.
sources <- list(
  list(file = "shapefiles/Barents Sea/atlantic_water.shp",    type = "Atlantic", region = "Barents Sea"),
  list(file = "shapefiles/Barents Sea/arctic_water.shp",      type = "Arctic",   region = "Barents Sea"),
  list(file = "shapefiles/North Atlantic/warm_currents_v2.shp", type = "Atlantic", region = "North Atlantic"),
  list(file = "shapefiles/North Atlantic/cold_currents_v2.shp", type = "Arctic",   region = "North Atlantic")
)

read_source <- function(file, type, region) {
  x <- sf::st_transform(sf::st_read(file, quiet = TRUE), 4326)
  co <- as.data.frame(sf::st_coordinates(x))      # X, Y, L1 (feature index)
  tag <- substr(region, 1, regexpr(" ", paste0(region, " ")) - 1)
  data.frame(
    long   = round(co$X, 6),
    lat    = round(co$Y, 6),
    order  = ave(co$L1, co$L1, FUN = seq_along),
    group  = paste(type, tag, co$L1, sep = "_"),
    size   = as.integer(x$size[co$L1]),
    type   = type,
    region = region,
    stringsAsFactors = FALSE
  )
}

all_parts <- lapply(sources, function(s) read_source(s$file, s$type, s$region))
all_currents <- do.call(rbind, all_parts)

dir.create("tabular", showWarnings = FALSE)

na <- all_currents[all_currents$region == "North Atlantic", ]
write.csv(na, "tabular/north_atlantic_currents.csv", row.names = FALSE)
write.csv(all_currents, "tabular/all_currents.csv", row.names = FALSE)

message(sprintf("north_atlantic_currents.csv: %d arrows, %d vertices",
                length(unique(na$group)), nrow(na)))
message(sprintf("all_currents.csv: %d arrows, %d vertices",
                length(unique(all_currents$group)), nrow(all_currents)))

File: python/produce_csv.py
"""Produce tidy CSV files of ocean-current arrows from the source shapefiles.

This is the modern, reproducible replacement for the historical
``Produce csv.R`` script. It reads the curated current-arrow shapefiles,
reprojects everything to WGS84 decimal degrees (EPSG:4326), and writes long-
format CSV tables that are easy to plot in R (ggOceanMaps), Python
(matplotlib/geopandas), or any GIS/Matlab workflow.

Usage
-----
    python python/produce_csv.py

Outputs (written to ``tabular/``)
---------------------------------
- ``north_atlantic_currents.csv`` : North Atlantic / Arctic Ocean arrows.
- ``all_currents.csv``            : Barents Sea + North Atlantic combined.

Schema (long format, one row per vertex)
----------------------------------------
- ``long``, ``lat`` : vertex coordinates in decimal degrees (EPSG:4326).
- ``order``         : vertex order within an arrow (1 = start, n = arrow head).
- ``group``         : unique arrow id (``<type>_<region>_<n>``).
- ``size``          : relative current strength (1 = weak ... 5 = strong).
- ``type``          : water-mass origin, ``"Atlantic"`` (warm) or ``"Arctic"`` (cold).
- ``region``        : ``"Barents Sea"`` or ``"North Atlantic"``.

Dependencies: geopandas, pandas (see requirements.txt).
"""

from __future__ import annotations

import os
import sys

import geopandas as gpd
import pandas as pd

WGS84 = "EPSG:4326"
HERE = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
SHP = os.path.join(HERE, "shapefiles")
OUT = os.path.join(HERE, "tabular")

# Source shapefiles -> (type, region). "warm" currents carry Atlantic water,
# "cold" currents carry Arctic water; we harmonise to the Barents Sea
# Atlantic/Arctic convention so a single colour scale works across regions.
# The North Atlantic files (EPSG:3995, Arctic Polar Stereographic) span the
# subtropical Atlantic through the Arctic Ocean; we reproject them to decimal
# degrees for the tidy tables.
SOURCES = {
    "Barents Sea": [
        (os.path.join(SHP, "Barents Sea", "atlantic_water.shp"), "Atlantic"),
        (os.path.join(SHP, "Barents Sea", "arctic_water.shp"), "Arctic"),
    ],
    "North Atlantic": [
        (os.path.join(SHP, "North Atlantic", "warm_currents.shp"), "Atlantic"),
        (os.path.join(SHP, "North Atlantic", "cold_currents.shp"), "Arctic"),
    ],
}


def line_to_rows(geom, group, size, type_, region):
    """Explode a (multi)line geometry into ordered vertex rows."""
    if geom.geom_type == "MultiLineString":
        coords = [pt for part in geom.geoms for pt in part.coords]
    else:
        coords = list(geom.coords)
    return [
        {
            "long": round(x, 6),
            "lat": round(y, 6),
            "order": i + 1,
            "group": group,
            "size": int(size),
            "type": type_,
            "region": region,
        }
        for i, (x, y, *_) in enumerate(coords)
    ]


def build_region(region, sources):
    rows = []
    for path, type_ in sources:
        if not os.path.exists(path):
            sys.exit(f"Missing source shapefile: {path}")
        gdf = gpd.read_file(path).to_crs(WGS84)
        for n, (_, feature) in enumerate(gdf.iterrows(), start=1):
            size = feature.get("size", 3)
            group = f"{type_}_{region.split()[0]}_{n}"
            rows.extend(line_to_rows(feature.geometry, group, size, type_, region))
    return pd.DataFrame(rows)


def main():
    os.makedirs(OUT, exist_ok=True)

    na = build_region("North Atlantic", SOURCES["North Atlantic"])
    na.to_csv(os.path.join(OUT, "north_atlantic_currents.csv"), index=False)
    print(f"north_atlantic_currents.csv: {na.group.nunique()} arrows, {len(na)} vertices")

    parts = [build_region(r, s) for r, s in SOURCES.items()]
    allc = pd.concat(parts, ignore_index=True)
    allc.to_csv(os.path.join(OUT, "all_currents.csv"), index=False)
    print(
        f"all_currents.csv: {allc.group.nunique()} arrows, {len(allc)} vertices "
        f"({', '.join(sorted(allc.region.unique()))})"
    )


if __name__ == "__main__":
    main()
